Replace ESMF subprocess calls with new parallel implementation of xesmf (0.8.x) directly in Python (#73)
COSIMA/regional-mom6: README.md

3. A bathymetry file that at least covers your domain
4. 3D ocean forcing files *of any resolution* on your choice of A, B or C Arakawa grid
5. Surface forcing files (eg ERA or JRA reanalysis)

> 6. A copy of [GFDL's FRE tools](https://github.com/NOAA-GFDL/FRE-NCtools) downloaded and compiled somewhere.

Reviewer: ```suggestion
6. [GFDL's FRE tools](https://github.com/NOAA-GFDL/FRE-NCtools) must be downloaded and compiled somewhere.
```

Author: ```suggestion
6. [GFDL's FRE tools](https://github.com/NOAA-GFDL/FRE-NCtools) must be downloaded and compiled on the machine you are using.
```

## Getting started

The two example notebooks walk you through how to use the package on two different sets of input datasets. Make sure that you can get at least one of these working on your setup with your MOM6 executable, and then try to modify them to apply to your domain with your bathymethry / forcing / boundaries. 

### Note for larger domains

> The example notebooks deal with a pretty small domain, but might not scale well if you try to run something really computationally expensive. When we first wrote this tool, xesmf (the package that handles all the regridding) didn't support parallelised weight generation. This meant that for large domains, you needed to call the underlying ESMF library with `mpirun`. This is especially important for the `.bathymetry` method, as the weight generation step becomes very slow for large areas if there's no spatial chunking. With the [introduction of `xesmf 0.8.x`](https://xesmf.readthedocs.io/en/latest/changes.html), a parallel option is now possible directly in python. 

Reviewer: if we require xesmf v0.8 or later, can't we add that compat requirement in the package dependencies? @angus-g?

Reviewer: will it be something that has to do with this line?
https://github.com/COSIMA/regional-mom6/blob/d9b997f0e497d58db4a827432be7c88d2b472b1b/pyproject.toml#L16

Reviewer: Done

> The example notebooks deal with a pretty small domain, but might not scale well if you try to run something really computationally expensive. When we first wrote this tool, xesmf (the package that handles all the regridding) didn't support parallelised weight generation. This meant that for large domains, you needed to call the underlying ESMF library with `mpirun`. This is especially important for the `.bathymetry` method, as the weight generation step becomes very slow for large areas if there's no spatial chunking. With the [introduction of `xesmf 0.8.x`](https://xesmf.readthedocs.io/en/latest/changes.html), a parallel option is now possible directly in python. 

Reviewer: ```suggestion
The example notebooks deal with a pretty small domains; they might not scale well if you try to run something really computationally expensive. When we started developing this package, xesmf (the package that handles all the regridding) didn't support parallelised weight generation. This meant that for large domains, users needed to call the underlying ESMF library with `mpirun`. This is especially important for the `.bathymetry` method, as the weight generation step becomes very slow for large domains if there's no spatial chunking. With the [introduction of `xesmf 0.8.x`](https://xesmf.readthedocs.io/en/latest/changes.html), a parallel option is now possible directly in python. 
```

Reviewer: Also, do we need to give people the remark about the history of the development here? Can we, perhaps, avoid all the discussion since now our package requires xesmf v0.8 or later so that's sorted, right?

> The example notebooks deal with a pretty small domain, but might not scale well if you try to run something really computationally expensive. When we first wrote this tool, xesmf (the package that handles all the regridding) didn't support parallelised weight generation. This meant that for large domains, you needed to call the underlying ESMF library with `mpirun`. This is especially important for the `.bathymetry` method, as the weight generation step becomes very slow for large areas if there's no spatial chunking. With the [introduction of `xesmf 0.8.x`](https://xesmf.readthedocs.io/en/latest/changes.html), a parallel option is now possible directly in python. 

Reviewer: I think a comment is needed here explaining what is "small" and what is "large" domain. (order of magnitude...?)

Reviewer: Also, do we need to give people the remark about the history of the development here? Can we, perhaps, avoid all the discussion since now our package requires xesmf v0.8 or later so that's sorted, right?

Reviewer 2: Agreed

Author: Yeah agreed - we could remove all of this for now and revise the readme in a later PR

> 
> ### Note for larger domains
> 
> The example notebooks deal with a pretty small domain, but might not scale well if you try to run something really computationally expensive. When we first wrote this tool, xesmf (the package that handles all the regridding) didn't support parallelised weight generation. This meant that for large domains, you needed to call the underlying ESMF library with `mpirun`. This is especially important for the `.bathymetry` method, as the weight generation step becomes very slow for large areas if there's no spatial chunking. With the [introduction of `xesmf 0.8.x`](https://xesmf.readthedocs.io/en/latest/changes.html), a parallel option is now possible directly in python. 

Author: ```suggestion
```

> If you have the latest version of `xesmf` the package will attempt to regrid in parallel, and if not will throw a warning and run it in serial. It will also print out the relevant `mpirun` command which you could use as a backup. Depending on your setup, you may need to write scripts that implement the package to access more computational resources than might be available in a Jupyter notebook as is the case at the NCI facility. 

Reviewer: No need to say "if" since now by requirement users _will have to have_ xesmf v0.8 or later.

Also, remove anything related with NCI from here.

Reviewer: NCI-related comments belong in the example at https://github.com/COSIMA/cosima-recipes/pull/216

Reviewer: @ashjbarnes we should deal with this also; please remove all NCI refs from README?

Reviewer: f09f566e88fa95360b395842041f453cde04c2ae deals with this
